packages/components/src/internal/popup-utils.ts

```diff
@@ -0,0 +1,173 @@
+    }
+  };
+
+  const position = getPosition(preferredSide);
+
+  // If the popup would overflow the viewport when placed on the preferred side of the target
+  // element, "flip" it to the other side of the element.
+  const maybeFlip = (d: Side): Side => {
+    switch (d) {
+      case 'top':
+        return position.top < 0 ? 'bottom' : 'top';
+      case 'right':
+        return position.left + elementSize.width >= viewportSize.width ? 'left' : 'right';
+      case 'bottom':
+        return position.top + elementSize.height >= viewportSize.height ? 'top' : 'bottom';
+      case 'left':
+        return position.left < 0 ? 'right' : 'left';
+      default:
+        throw new Error('Invalid side');
+    }
+  };
+
+  const finalSide = maybeFlip(preferredSide);
+
+  // Recompute position after side may have changed.
+  const finalPosition = getPosition(finalSide);
+
+  return {
+    position: {
+      ...finalPosition,
+      right: finalPosition.left + elementSize.width,
+      bottom: finalPosition.top + elementSize.height,
+    },
+    side: finalSide,
+  };
+};
+
+type BoundingRect = Size & Position;
+
+const toBoundingRect: (domRect: DOMRect) => BoundingRect = _.pick([
+  'width',
+  'height',
+  'top',
+  'right',
+  'bottom',
+  'left',
+]);
+
+export type UseBoundingRectsSelector =
+  | { ref: RefObject<HTMLElement | null | undefined> }
+  | { id: string }
+  | { viewport: true };
+
+/**
+ * Returns the bounding rectangles of specified elements.
+ * @param selectors - Selectors for elements. Selectors can be by ref, ID, or the viewport/window.
+ */
+export const useBoundingRects = (selectors: UseBoundingRectsSelector[]): BoundingRect[] => {
+  const [dimensions, setDimensions] = useState<BoundingRect[]>(
+    selectors.map(() => ({ width: 0, height: 0, top: 0, right: 0, bottom: 0, left: 0 }))
+  );
+
+  const dimensionsRef = useRef<BoundingRect[]>();
+  dimensionsRef.current = dimensions;
+
+  const animationRef = useRef<number>();
+
+  const computePosition = useCallback(() => {
+    const newDimensions: BoundingRect[] = selectors.map((selector) => {
+      if ('ref' in selector && selector.ref.current) {
+        return toBoundingRect(selector.ref.current.getBoundingClientRect());
+      }
+
+      if ('id' in selector) {
+        const element = document.getElementById(selector.id);
+        if (element) {
+          return toBoundingRect(element.getBoundingClientRect());
+        }
+      }
+
+      if ('viewport' in selector) {
+        const width = window.innerWidth;
+        const height = window.innerHeight;
+        return { width, height, top: 0, right: width, bottom: height, left: 0 };
+      }
+
+      return { width: 0, height: 0, top: 0, right: 0, bottom: 0, left: 0 };
+    });
+
+    if (!_.isEqual(newDimensions, dimensionsRef.current)) {
+      setDimensions(newDimensions);
+    }
+
+    animationRef.current = requestAnimationFrame(computePosition);
+  }, [selectors]);
+
+  useEffect(() => {
```

Comment:
Why `useEffect` instead of `usOnMount`?

Reply:
I hope to get rid of `useOnMount`. I find that `useOnMount` almost always results in a bug as the effect is no longer re-run when its dependencies change.

Reply:
For example, using `useOnMount` here would result in this hook always returning bounding rects for its initial list of selectors, even if the consumer calls it with different selectors on a later render.

Reply:
And just like that, you identified the cause of https://broadworkbench.atlassian.net/browse/UIE-125.

Can we get the terra-ui version of popup-utils to this one? If not, we should make the `useEffect` fix there.

Reply:
Updated this in https://github.com/DataBiosphere/terra-ui/pull/4234/commits/c02a64bd86c91c531b25ac0da1c1375cf39ffee8, so `useOnMount` would actually work here. But I still think that `useOnMount` is generally harmful and being forced to detail with the dependencies of `useEffect` leads to more correct behavior.

Reply:
> Can we get the terra-ui version of popup-utils to this one? If not, we should make the useEffect fix there.

After these outstanding components PRs get merged, my next step is to move TooltipTrigger to components where it will use this version.